base/bits.hpp

```diff
@@ -89,132 +46,107 @@ namespace bits
+  x = ((x & 0x22222222) << 1) | ((x >> 1) & 0x22222222) | (x & 0x99999999);
+  return x;
+}
+
+inline constexpr uint32_t PerfectUnshuffle(uint32_t x)
+{
+  x = ((x & 0x22222222) << 1) | ((x >> 1) & 0x22222222) | (x & 0x99999999);
+  x = ((x & 0x0C0C0C0C) << 2) | ((x >> 2) & 0x0C0C0C0C) | (x & 0xC3C3C3C3);
+  x = ((x & 0x00F000F0) << 4) | ((x >> 4) & 0x00F000F0) | (x & 0xF00FF00F);
+  x = ((x & 0x0000FF00) << 8) | ((x >> 8) & 0x0000FF00) | (x & 0xFF0000FF);
+  return x;
+}
+
+// Returns the integer that has the bits of |x| at even-numbered positions
+// and the bits of |y| at odd-numbered positions without changing the
+// relative order of bits coming from |x| and |y|.
+// That is, if the bits of |x| are {x31, x30, ..., x0},
+//         and the bits of |y| are {y31, y30, ..., y0},
+// then the bits of the result are {y31, x31, y30, x30, ..., y0, x0}.
+inline constexpr uint64_t BitwiseMerge(uint32_t x, uint32_t y)
+{
+  uint32_t const hi = PerfectShuffle((y & 0xFFFF0000) | (x >> 16));
+  uint32_t const lo = PerfectShuffle(((y & 0xFFFF) << 16 ) | (x & 0xFFFF));
+  return (static_cast<uint64_t>(hi) << 32) + lo;
+}
+
+inline void BitwiseSplit(uint64_t v, uint32_t & x, uint32_t & y)
+{
+  uint32_t const hi = PerfectUnshuffle(static_cast<uint32_t>(v >> 32));
+  uint32_t const lo = PerfectUnshuffle(static_cast<uint32_t>(v & 0xFFFFFFFFULL));
+  x = ((hi & 0xFFFF) << 16) | (lo & 0xFFFF);
+  y = (hi & 0xFFFF0000) | (lo >> 16);
+}
+
+// Returns 1 if bit is set and 0 otherwise.
+inline uint8_t GetBit(void const * p, uint32_t offset)
+{
+  uint8_t const * pData = static_cast<uint8_t const *>(p);
+  return (pData[offset >> 3] >> (offset & 7)) & 1;
+}
+
+inline void SetBitTo0(void * p, uint32_t offset)
+{
+  uint8_t * pData = static_cast<uint8_t *>(p);
+  pData[offset >> 3] &= ~(1 << (offset & 7));
+}
+
+inline void SetBitTo1(void * p, uint32_t offset)
+{
+  uint8_t * pData = static_cast<uint8_t *>(p);
+  pData[offset >> 3] |= (1 << (offset & 7));
+}
   
-  // Computes number of bits needed to store the number, it is not equal to number of ones.
-  // E.g. if we have a number (in bit representation) 00001000b then NumUsedBits is 4.
-  inline uint32_t NumUsedBits(uint64_t n)
-  {
-    uint32_t result = 0;
-    while (n != 0) { ++result; n >>= 1; }
-    return result;
-  }
-
-  inline uint64_t GetFullMask(uint8_t numBits)
-  {
-    ASSERT_LESS_OR_EQUAL(numBits, 64, ());
-    return numBits == 64 ? std::numeric_limits<uint64_t>::max()
-                         : (static_cast<uint64_t>(1) << numBits) - 1;
-  }
-
-  inline bool IsPow2Minus1(uint64_t n) { return (n & (n + 1)) == 0; }
+// Computes number of bits needed to store the number, it is not equal to number of ones.
+// E.g. if we have a number (in bit representation) 00001000b then NumUsedBits is 4.
+inline uint32_t NumUsedBits(uint64_t n)
+{
+  uint32_t result = 0;
+  while (n != 0) { ++result; n >>= 1; }
+  return result;
+}
+
+inline constexpr uint64_t GetFullMask(uint8_t numBits)
```

Comment:
`constexpr` functions are implicitely `inline`, so you can just keep the `constexpr` kw.